runtime/near-vm-logic/src/logic.rs

```diff
@@ -696,6 +696,23 @@ impl<'a> VMLogic<'a> {
         Ok(self.current_storage_usage)
     }
 
+    /// Returns the block hash for the block at the given height (if any). It can
+    /// only return a result for the most recent 256 blocks.
+    ///
+    /// # Cost
+    ///
+    /// `base + storage_read_base + storage_read_key_byte * 8 + write_register_base + write_register_byte * 32
+    #[cfg(feature = "protocol_feature_block_hash_host_fn")]
+    pub fn block_hash(&mut self, block_height: u64, register_id: u64) -> Result<()> {
+        self.gas_counter.pay_base(base)?;
+        // we assume that under the hood this is a DB read, where the block height
+        // is the key, and it is 64 bits = 8 bytes
+        self.gas_counter.pay_base(storage_read_base)?;
+        self.gas_counter.pay_per_byte(storage_read_key_byte, 8)?;
+        let result = self.ext.block_hash(block_height)?.unwrap_or_default();
```

Comment:
Hm, is this the best ABI possible? My understanding is that the hash might, or might not be there, so essentailly we are returning an `Option<Vec<u8>>`. But we encode optionnness in the length of the vector, so the user needs to explicitly check for it. 

I guess, at minimum we need to explicitly document "if the block's hash is unavailble, the returned vector has zero length". But maybe a more direct API is possible? Could we return a `bool` here, and only write a vector if the hash actually exists?

Reply:
It's not returning a vector of 0 length. It always returns a `[u8; 32]` (i.e. 256-bit sequence), but in the case where the hash is missing it returns `[0u8; 32]`. This is still perhaps not the best way to do this because the consumer of the API still has to check that the result is non-zero. I like the suggestion of returning a `bool` and only writing to the register if the hash is present!

Reply:
Done in https://github.com/near/nearcore/pull/4273/commits/401d95197344e4f28695d3283603911938bda108